jdaviz/app.py

```diff
@@ -911,36 +913,84 @@ def add_data(self, data, data_label=None, notify_done=True):
+            For example, "filename[FLUX]" where "FLUX" is the ext value.
+        alt_name : str, optional
+            Alternate names that can be used if none of the options provided are valid.
+        check_unique : bool
+            Included so that this method can be used with data label retrieval in addition
+            to generation.
+
         Returns
         -------
-        A string data label used by Glue
+        data_label : str
+            A unique data label that at its root is either given by the user at load time
+            or created by jdaviz using a description of the loaded file's type.
+        """
+        data_label = None
+
+        if loaded_object is None:
+            pass
+        elif isinstance(loaded_object, str):
+            # This is either the full file path or the basename or the user input data label
+            if loaded_object.lower().endswith(('.jpg', '.jpeg', '.png')):
+                file_format = loaded_object.lower().split(".")[-1]
+                loaded_object = os.path.splitext(os.path.basename(loaded_object))[0]
+                data_label = f"{loaded_object}[{file_format}]"
+            elif not os.path.isfile(loaded_object) or re.search(r'(.+)(\[(.*?)\])$',
+                                                                loaded_object):
+                # Not a file path so assumed to be a data label
+                data_label = f"{loaded_object}"
+            else:
+                loaded_object = os.path.splitext(os.path.basename(loaded_object))[0]
+                data_label = f"{loaded_object}"
+        elif isinstance(loaded_object, fits.hdu.hdulist.HDUList):
+            if hasattr(loaded_object, 'file_name'):
+                data_label = f"{loaded_object.file_name}[HDU object]"
+            else:
+                data_label = "Unknown HDU object"
+        elif isinstance(loaded_object, Spectrum1D):
+            data_label = "Spectrum1D"
+        elif isinstance(loaded_object, NDData):
+            data_label = "NDData"
+        elif isinstance(loaded_object, np.ndarray):
+            data_label = "Array"
 
-        """
+        if data_label is None and alt_name is not None and isinstance(alt_name, str):
+            data_label = alt_name
+        elif data_label is None:
+            data_label = "Unknown"
 
-        # if path is not a file or already ends in [ ] extension, assume it is a data-label
-        if not os.path.isfile(path) or re.search(r'(.+)(\[(.*?)\])$', path):
-            return path
-        else:
-            assert ext, 'A data extension must be specified'
-            p = pathlib.Path(path)
-            stem = p.stem.split(os.extsep)[0]
-            label = f'{stem}[{ext}]'
-            return label
+        if ext is not None:
+            data_label = data_label + f"[{ext}]"
+
+        if check_unique:
+            data_label = self.return_unique_name(data_label)
+
+        return data_label
+
+    def return_unique_name(self, data_label):
```

Comment:
Would be nice to have a unit test to test just this method with different cases (without actually loading the data).

Also we have to think of the implications of this for when we start to let people delete loaded data.

Reply:
Good call, tests have been added.